coqpyt/coq/proof_file.py

```diff
@@ -425,30 +443,56 @@ def __is_proof_term(self, step: Step):
         return self.context.expr(step)[0] in ["VernacEndProof", "VernacExactProof"]
 
-    def __check_proof_step(self, step: Step, goals: Optional[GoalAnswer] = None):
+    def __check_proof_step(
+        self, step: Step, goals: Optional[GoalAnswer], undo: bool = False
+    ):
         # Last step of the file
         if step.text.strip() == "":
             return
 
-        # Found [Qed]/[Defined]/[Admitted] or [Proof <exact>.]
-        if self.__is_end_proof(step):
-            self.__handle_end_proof(step, goals)
-        elif self.__is_proof_term(step):
-            self.__handle_proof_term(step, goals)
+        if self.__is_proof_term(step):
+            self.__handle_proof_term(step, undo=undo)
         # Avoids Tactics, Notations, Inductive...
         elif self.context.term_type(step) == TermType.OTHER:
-            self.__handle_proof_step(step, goals)
+            self.__handle_proof_step(step, goals=goals, undo=undo)
+
+    def __is_segment_delimiter(self, step: Step):
+        return self.context.expr(step)[0] in [
+            "VernacEndSegment",
+            "VernacDefineModule",
+            "VernacDeclareModuleType",
+            "VernacBeginSection",
+        ]
 
-    def __forward_step(self, goals: GoalAnswer):
+    def __forward_step(self, goals: Optional[GoalAnswer]):
         self.__aux_file.append(self.prev_step.text)
-        self.__check_program()
-        if self.in_proof or self.__is_end_proof(self.prev_step):
+        # Found [Qed]/[Defined]/[Admitted] or [Proof <exact>.]
+        if self.__is_end_proof(self.prev_step):
+            self.__handle_end_proof(self.prev_step, goals=goals)
+            return
+        # Ignore segment delimiters because it affects Program handling
+        if self.__is_segment_delimiter(self.prev_step):
+            return
+        # Check if Program definition or proof step
+        is_program = self.__check_program(self.current_goals)
+        if not is_program and self.in_proof:
             self.__check_proof_step(self.prev_step, goals)
 
-    def __backward_step(self):
+    def __backward_step(self, goals: Optional[GoalAnswer]):
         self.__aux_file.truncate(self.curr_step.text)
-        if len(self.open_proofs) > 0 or self.__is_end_proof(self.curr_step):
-            self.__check_proof_step(self.curr_step)
+        # Found [Qed]/[Defined]/[Admitted] or [Proof <exact>.]
+        if self.__is_end_proof(self.curr_step):
+            self.__handle_end_proof(self.curr_step, undo=True)
+            return
+        # Ignore segment delimiters because it affects Program handling
+        if self.__is_segment_delimiter(self.curr_step):
+            return
+        # Check if Program definition or proof step
+        is_program = self.__check_program(
+            self.current_goals, prev_goals=goals, undo=True
+        )
+        if not is_program and len(self.open_proofs) > 0:
+            self.__check_proof_step(self.curr_step, goals, undo=True)
```

Comment:
I think we can merge these two functions. We just need a new function, for instance, called __step, that receives the goals, a step and undo.